### PR TITLE
fix(dnslink-gw): breadcrumbs and CID column when dir listing

### PR DESCRIPTION
**Pull request by lidel**
This PR aims to solve issue described in https://github.com/ipfs/dir-index-html/issues/42 
It changes URI of some broken links when directory listing occurs on DNSLink website adds a bunch of tests.

# Changes

There are no visual changes to the way directory listing is presented.

- [x] fixed breadcrumbs on dnslink website 
- [x] fixed CID column on dnslink website
  - fixing this will required chages in `dir-index-htm`, this PR is switching to a new release with fix from  https://github.com/ipfs/dir-index-html/pull/43
- [x] proper test coverage of dir listing (until now we were very lucki, as we had nearly no tests)
  - etag/cache control
  - path edge cases: `/api` /ipfs` and `ą/ś/ć` (utf8) 
  - breadcrumbs (path, subdomain, dnslink)
  - CID column (path, subdomain, dnslink)
- [x] <del>make everything work with `X-Ipfs-Gateway-Prefix`</del>
  - `X-Ipfs-Gateway-Prefix` is already mostly broken, filled https://github.com/ipfs/go-ipfs/issues/7702 to remove it, so not in the scope of this PR

# How to tests this PR

This PR mainly fixes the DNSLink mode, which is when domain passed in `Host` header has a valid DNSLink record:

```console
$ curl -H "Host: dist.ipfs.io" -sD - "http://127.0.0.1:8080/go-ipfs/v0.7.0"
```

It may be tricky to understand what is the fix. The best illustration is to look at CID column:
  - when path or subdomain gateway is used, CID column links to direct CID with filename parameter:
    ```console
    $ curl -sD - "http://127.0.0.1:8080/ipns/dist.ipfs.io/go-ipfs/v0.7.0" | grep "ipfs-hash"
    <a class="ipfs-hash" href="/ipfs/Qmf9JEWyETbLmAdyT9BaBStNeJ2ZHgvgXFNSWgdz9saa4N?filename=go-ipfs_v0.7.0_darwin-386.tar.gz">
    ```
  - when DNSLink gateway is used (`Host` header), CID column links to cid.ipfs.io:
    ```console
    $ curl -H "Host: dist.ipfs.io" -sD - "http://127.0.0.1:8280/go-ipfs/v0.7.0" | grep "ipfs-hash"
    <a class="ipfs-hash" href="https://cid.ipfs.io/#Qmf9JEWyETbLmAdyT9BaBStNeJ2ZHgvgXFNSWgdz9saa4N" target="_blank" rel="noreferrer noopener">
    ```

Similar fix is applied to breadcrumbs, but we link to domain from dnslink instead of a public gateway.

### CODE REVIEW

**Comment by jessicaschilling** [comment]
@lidel Apologies for the delay - v1.2.1 of dir-index-html is here: https://github.com/ipfs/dir-index-html/releases/tag/1.2.1

File: test/sharness/t0115-gateway-dir-listing.sh [hunk 1]
```diff
@@ -0,0 +1,146 @@
+#!/usr/bin/env bash
+#
+# Copyright (c) Protocol Labs
+
+test_description="Test directory listing (dir-index-html) on the HTTP gateway"
```
Comment (lidel) [inline test/sharness/t0115-gateway-dir-listing.sh:5]: This file adds proper end-to-end tests for directory listing on Gateway port that protects us against regressions such as https://github.com/ipfs/dir-index-html/issues/42 by testing each gateway type: 
- path gateway
- subdomain gateway
- dnslink website gateway
    
We check behavior of:
- etag/unicode support
- breadcrumbs ( 
- file name column
- hash column

This level of testing is necessary, because URIs will be different on each gw type, and we really don't want to introduce any regressions around website/directory hosting.

**Comment by lidel** [comment]
@aschmahmann I believe this is ready for your review.

**Review by aschmahmann — APPROVED** [review]
